xarray/backends/core.py

```diff
@@ -0,0 +1,425 @@
+import numpy as np
+import pickle
+import threading
+from xarray.core.pycompat import OrderedDict, dask_array_type
+
+from xarray import Variable
+
+
+class AbstractDecoder(object):
+    """Callable for decoding variables and attributes into an xarray.Dataset.
+
+    Because we need to put Decoders in tokens for use with dask, they should be
+    pickleable.
+    """
+
+    def __call__(self, variables, attrs):
+        # type: (Mapping[Any, xarray.Variable], Mapping) -> Dataset
+        """Convert variables and attributes into an xarray.Dataset.
+
+        Parameters
+        ----------
+        variables : Mapping[Any, xarray.Variable]
+            Mapping of xarray.Variable objects to use in the Dataset.
+        attrs : Mapping
+            Mapping of global attributes.
+
+        Returns
+        -------
+        xarray.Dataset with the decoded contents of ``variables`` and ``attrs``.
+        """
+        raise NotImplementedError
+
+
+class AbstractEncoder(object):
+    """Callable for decoding variables and attributes into an xarray.Dataset.
+
+    For use with dask, Encoders should be pickleable.
+    """
+
+    def __call__(self, dataset):
+        # type: (xarray.Dataset,) -> (Mapping[Any, xarray.Variable], Mapping)
+        """Convert an xarray.Dataset into maps of variables and attributes.
+
+        Recognized encoding options may be removed from
+        `variable.encodings` and used to encode the data (generally, they should
+        then be put into `variable.attrs`. Unrecognized encodings should be
+        passed on unchanged.
+
+        Parameters
+        ----------
+        dataset : xarray.Dataset
+            Dataset to encode.
+
+        Returns
+        -------
+        variables : Mapping[Any, xarray.Variable]
+            Mapping of variables to save into a DataStore. Should only have
+            dtypes supported by the DataStore.
+        attrs : Mapping
+            Mapping of global attributes for the DataStore.
+        """
+        raise NotImplementedError
+
+
+class DummyDecoder(AbstractDecoder):
+    def __call__(self, variables, attrs):
+        return xarray.Dataset(variables, attrs=attrs)
+
+
+class DummyEncoder(AbstractDecoder):
+    def __call__(self, dataset):
+        return dataset.variables, dataset.attrs
+
+
+# CFDecoder and CFEncoder would probably live in conventions.py:
+
+class CFDecoder(AbstractDecoder):
+    def __init__(self, concat_characters=True, mask_and_scale=True,
+                 decode_times=True, decode_endianness=True,
+                 drop_variables=None):
+        self._kwargs = {
+            'concat_characters': concat_characters,
+            'mask_and_scale': mask_and_scale,
+            'decode_times': decode_times,
+            'decode_endianness': decode_endianness,
+            'drop_variables': drop_variables,
+        }
+
+    def __call__(self, variables, attrs):
+        return conventions.decode_cf_variables(
+            variables, attrs, **self._kwargs)
```

Comment:
This looks great. It would definitely solve some of the encoding challenges with zarr.